8341566: Add Reader.of(CharSequence)

mkarg opened the pull request:
This Pull Requests proposes an implementation for [JDK-8341566](https://bugs.openjdk.org/browse/JDK-8341566): Adding the new method `public static Reader Reader.of(CharSequence)` will return an anonymous, non-synchronized implementation of a `Reader` for each kind of `CharSequence` implementation. It is optimized for `String`, `StringBuilder`, `StringBuffer` and `CharBuffer`.

In addition, this Pull Request proposes to replace the implementation of `StringReader` to become a simple synchronized wrapper around `Reader.of(CharSequence)` for the case of `String` sources. To ensure correctness, this PR...
* ...simply moved the **original code** of `StringBuilder` to become the de-facto implementation of `Reader.of()`, then stripped synchronized from it on the left hand, but kept just a synchronized wrapper on the right hand. Then added a `switch` for optimizations within the original code, at the exact location where previously just an optimization for `String` lived in.
* ...added tests for all methods (`Of.java`), and applied that test upon the modified `StringBuilder`.

Wherever new JavaDocs were added, existing phrases from other code locations have been copied and adapted, to best match the same wording.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8341596](https://bugs.openjdk.org/browse/JDK-8341596) to be approved
- [x] Change must be properly reviewed (2 reviews required, with at least 2 [Reviewers](https://openjdk.org/bylaws#reviewer))

### Issues
 * [JDK-8341566](https://bugs.openjdk.org/browse/JDK-8341566): Add Reader.of(CharSequence) (**Enhancement** - P4)
 * [JDK-8341596](https://bugs.openjdk.org/browse/JDK-8341596): Add Reader.of(CharSequence) (**CSR**)


### Reviewers
 * [Roger Riggs](https://openjdk.org/census#rriggs) (@RogerRiggs - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)
 * [Chen Liang](https://openjdk.org/census#liach) (@liach - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21371/head:pull/21371` \
`$ git checkout pull/21371`

Update a local copy of the PR: \
`$ git checkout pull/21371` \
`$ git pull https://git.openjdk.org/jdk.git pull/21371/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21371`

View PR using the GUI difftool: \
`$ git pr show -t 21371`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21371.diff">https://git.openjdk.org/jdk/pull/21371.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21371#issuecomment-2395111947)

mkarg commented on test/jdk/java/io/Reader/Of.java:
```diff
@@ -0,0 +1,199 @@
+    @DataProvider
+    public static Reader[] readers() {
+        return new Reader[] {
+            new StringReader(CONTENT),
```
Explicitly including that original class here (even if it has nothing to do with the `of` method) to be sure that we did not modify it in an incompatible way. Unfortunately there is no full test coverage for `StringReader`, and it does not make much sense to duplicate the tests.

mkarg replied:
Good idea. But instead, we could also add a new API `CharSequence.of(char[])` and test *that* class? WDYT?

ecki replied:
> Good idea. But instead, we could also add a new API `CharSequence.of(char[])` and test _that_ class? WDYT?

Both you need to test the generic case anyway.

mkarg commented on src/java.base/share/classes/java/io/StringReader.java:
```diff
@@ -207,7 +177,11 @@ public void reset() throws IOException {
+            try {
+                r.close();
+            } catch (IOException e) {
+                throw new UncheckedIOException(e);
```
In fact, the implementation of `close` in the original class will never throw `IOException`, but unfortunately `Reader.close()` declares `throws IOException`.

mkarg commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,208 @@ public void close() {
+            @Override
+            public void close() {
+                cs = null;
+            }
```
@AlanBateman I need to confess that I did not understand what you had in mind when you wrote this on the mailing list:
> That doesn't excuse you completely from thinking about concurrent use as Readers have a close method so you'll need to think about how close is specified for when it is called while another thread is reading chars from a custom CS.

As this new implementation explicitly is for single-threaded use only, there is no such other thread that could call `close` concurrently.

Maybe I am missing something here, so I would kindly ask for an outline of a scenario where -despite the explicit single-thread note- a second thread *does* exist?

liach replied:
A scenario would be that this reader is shared by multiple threads in an argument to `Runnable`, like create 100 runnables, 50 are trying to close while 50 are reading to char arrays. There will be scenarios where some reading thread is reading while some closing thread invokes `close` and completes before the reading thread completes.

I think the key takeaway here is that `close` does not block until other threads finish reading, does not affect already-begun reading operations, and the closed state might not be immediately observed by other threads (as we do not release-write and acquire-read the cs clearing state). However, I don't know which of these attributes should be explicitly specified in our API specification.

mkarg replied:
As Alan suggested, I just now in this very minute have added a more clear statement about the fact that it is invalid to share this Reader instance among multiple threads without *external* synchronization. That will prevent the scenario you mention.

AlanBateman commented:
/csr

AlanBateman commented:
Reader.of(CharSequence) looks much better than introducing CharSequenceReader. It won't have an equivalent on Writer but I think that is okay. Also it means that the user will need to deal with close throwing IOException but anything using Reader has to do this already.

I think it would be better to drop "API compatible with StringReader" from the method description. An apiNote in StringReader can direct readers to the static factory method.

Also I think drop the  "lock" field as it's a protected field and only interesting to subclasses.  The Reader class does not specify if Reader is thread-safe so the method description doesn't need to say too much. For clarify then it could just say something like "the resulting Reader is not safe for use by multiple concurrent threads. If the Reader is to be used by more than one thread it should be controlled by appropriate synchronization".

The parameter name is currently "c", maybe you mean "cs"? The method will need to specify NPE for the of(null) case.

liach commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,208 @@ public void close() {
+     *
+     * @since 24
+     */
+    public static Reader of(CharSequence c) {
```
Should we give this factory a more specific name so we don't clash in the future? For example, if we add another factory `of(A a)` for interface `A`, then it would be confusing to have an instance of `interface C extends CharSequence, A` to be passed to `of`.

mkarg replied:
I do not see what is "confusing" in that case, as the caller still gets what he intends: a reader for the passed source. I also wonder how likely that case actually is. Anyways, I may be biased as I proposed `of`.

@AlanBateman WDYT?

AlanBateman replied:
You'll get different opinions on naming, personally I think Reader.of(cs) is very readable at use-sites. There are several other static methods named "of" added in other areas in recent years.

mkarg replied:
So let's go with `Reader.of()` unless nobody has stronger feelings against it. 😃

liach commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,208 @@ public void close() {
+     * use as a drop-in replacement for {@code StringReader} in places where the
+     * reader was being used by a single thread (as is generally the case).
+     * Where possible, it is recommended that this class be used in preference
+     * to {@code StringReader} as it will be faster under most implementations.
```
I believe this paragraph is redundant, now that `StringReader` is a somewhat-deprecated API that we aim to migrate away from. We can note these differences in the API notes of `StringReader`.

mkarg replied:
As Alan already proposed, I am just removing it in this very minute.

liach commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,208 @@ public void close() {
+     * {@code read(CharBuffer)}, {@code ready()}, {@code skip(long)}, and
+     * {@code transferTo()} methods all throw {@code IOException}.
+     *
+     * <p> The {@code markSupported()} method returns {@code true}.
```
```suggestion
     * <p>The returned reader supports the {@link #mark mark()} operation.
```

mkarg replied:
This is not wrong. This style is used all over the place already within `Reader.java`, so it would be strange to change it *just here*.

ecki replied:
You can “elaborately” the close and mark protocol in the Reader interface and then only mention itnadgeres to it here.

liach commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,208 @@ public void close() {
+     * <p> After the stream has been closed, the {@code read()},
+     * {@code read(char[])}, {@code read(char[], int, int)},
+     * {@code read(CharBuffer)}, {@code ready()}, {@code skip(long)}, and
+     * {@code transferTo()} methods all throw {@code IOException}.
```
I believe these 2 paragraphs are all redundant besides the initial `The returned stream is initially open.` sentence: those are duplicating information already available from `Reader::close` specification.

Closing `StringReader` is optional but its specification doesn't mention this. I think we don't need to mention if closing is optional, so that one sentence is sufficient.

mkarg replied:
This text already exists in this very file, it is a 1:1 copy from `nullReader()`. It cannot be wrong for `of()` if it was correct for `nullReader()`. To stay consistent with existing JavaDocs in this very same file, we should not have different texts for the same cases.

liach replied:
`nullReader` needs this because it specifies the open behavior of these methods. This API does not have to; we can check out how others think in the CSR review.

mkarg replied:
I have forwarded the current proposal as-is to the CSR.

mkarg commented:
Thank you, Alan! I have updated the JavaDocs according to your comments:
* Dropped "API compatible with StringReader" from description
* @apiNote in StringReader refers to static factory method
* Dropped "lock" field from API docs
* Added "The resulting Reader is not safe for use by multiple concurrent threads. If the Reader is to be used by more than one thread it should be controlled by appropriate synchronization"
* Deviating from your proposal, I renamed parameter "c" to "source" for clarity as the name "cs" already exists as an internal variable
* Method specifies NPE for `of(null)` case

I addition, JavaDocs now says "reader", not "stream" anymore.

mkarg commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,205 @@ public void close() {
+
+        return new Reader() {
+            private final int length = source.length();
+            private CharSequence cs = source;
```
This is correct, but then we would keep the reference to the char sequence "forever", without any good reason. This content could be *huge* (and often it is, as `CharSequence` most often is a `StringBuilder` just because the application *wants* to create huge content, and `String`-concatenation would be too costly with *huge* content). In fact, I see this scenario as the top use case for this new API.

Having said that, I would really like to have this field non-final to unlink the reference to the source as soon as possible, implicitly allowing GC to free this unused memory.

liach replied:
> keep the reference to the char sequence "forever"

Note that this only happens if the `Reader` instance is kept forever; but a `Reader` is supposed to be used and closed and discarded instead of being held indefinitely after closing. If users are bad enough, they can never call `close` and the non-final `cs` is still leaked.

The right way of GC freeing is to let the GC free the reader and the char sequence together, instead of allowing to free when a client keeps this reader instance erroneously after closing.

mkarg replied:
I really do see you point, nevertheless I think we should be kind to "bad" users, too. In the end, there is *no* specification which explicitly tells "bad" users that readers are support to be short-living (despite it being rather obvious), while `close` explicitly mentions freeing *all* resources, so it is rather unexpected to keep something past `close`, even for short term. NB: `StringReader` applies the very same behavior, too (using `s` and `str` variables), that's where I copied the code from originally. It is not a big problem for us to have two variables, neither that the name of the variable is `source` (in fact, I do like `source` even more than `cs`, but this is just my personal preference). To sum up, I really dislike the idea to keep the reference for any longer than essentially needed. If it's just about the name of the variable, I could simply switch them if you prefer that.

liach replied:
On a linux x64 build:
```
./javap -p java.io.Reader\$2
Compiled from "Reader.java"
class java.io.Reader$2 extends java.io.Reader {
  private final int length;
  private java.lang.CharSequence cs;
  private int next;
  private int mark;
  final java.lang.CharSequence val$source;
  java.io.Reader$2(java.lang.CharSequence);
  private void ensureOpen() throws java.io.IOException;
  public int read() throws java.io.IOException;
  public int read(char[], int, int) throws java.io.IOException;
  public long skip(long) throws java.io.IOException;
  public boolean ready() throws java.io.IOException;
  public boolean markSupported();
  public void mark(int) throws java.io.IOException;
  public void reset() throws java.io.IOException;
  public void close();
}
```

The javap in the built binaries of this patch shows that this anonymous class already keeps the `CharSequence val$source` in a final field; therefore, your setting the `cs` field to `null` will not help GC or anything of that sort.

mkarg commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,205 @@ public void close() {
+            /**
+             * Reads a single character.
+             *
+             * @return     The character read, or -1 if the end of the stream has been
```
Correct. If everybody is fine with that, I will drop it.

mkarg replied:
Dropped it.

mkarg commented:
> /csr

@AlanBateman Can you please review [the CSR request](https://bugs.openjdk.org/browse/JDK-8341596) so I can finish it? Thanks!

ecki commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,124 @@ public void close() {
+            @Override
+            public boolean ready() throws IOException {
+                ensureOpen();
+                return true;
```
I guess it’s not wrong, but can this be false when end of source is reached? Related to this you already described it is not synchronized but should we also describe it does not support changes to the length during “iteration” (could be in the same thread).

ecki replied:
Ignore my “thinko”/comment. of course it needs to return true for the EOD.

AlanBateman commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,124 @@ public void close() {
     }
 
+    /**
+     * Returns a new {@code Reader} whose source is a {@link CharSequence}.
```
Probably better to more this more precise, e.g.  Returns a {@code Reader} that reads characters from a {@code CharSequence}, starting at the first character in the sequence.

ecki commented on test/jdk/java/io/Reader/Of.java:
```diff
@@ -0,0 +1,199 @@
+
+    @Test(dataProvider = "readers")
+    public void testOpen(Reader reader) {
+        assertNotNull(reader, "Reader.of(String) returned null");
```
This does not look like a useful test and the assert is misleading (no String)

AlanBateman commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,124 @@ public void close() {
+     * calling the {@code close()} method. Subsequent calls to {@code close()}
+     * have no effect.
+     *
+     * <p> After the reader has been closed, the {@code read()},
```
This paragraph is okay but the Reader methods shouldn't specify this (doing so after 25+ years of being unspecified would require being confident wouldn't invalidate any existing Readers).

AlanBateman commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,124 @@ public void close() {
+     * {@code read(CharBuffer)}, {@code ready()}, {@code skip(long)}, and
+     * {@code transferTo()} methods all throw {@code IOException}.
+     *
+     * <p> The {@code markSupported()} method returns {@code true}.
```
Chen's suggestion to say that it supports the mark operation would be clearer than the legacy wording.

ecki commented on test/jdk/java/io/Reader/Of.java:
```diff
@@ -0,0 +1,199 @@
+        StringWriter sw = new StringWriter(16);
+        assertEquals(reader.transferTo(sw), 16, "transferTo() != 16");
+        assertEquals(reader.transferTo(sw), 0,
+                "transferTo() does not resect empty stream");
```
Typo respect (I also wonder if “empty stream” (used multiple times) should be “end of stream” or “exhausted stream”. The test never get an empty stream to test.

mkarg replied:
"empty stream" is actually correct. The case "end of stream" is implied in the line before (the test that expects result `16`, it starts with non-empty stream, it stops at end-of-stream). From that point, the stream is now empty, as `transferTo()` transferred *everything*. So what this test (the one expecting result `0`) tests is what happens if `transferTo()` starts in an already-empty situation.

bokken commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,124 @@ public void close() {
+                    return -1;
+                int n = Math.min(length - next, len);
+                switch (cs) {
+                    case String s -> s.getChars(next, next + n, cbuf, off);
```
There was some discussion on the mailing list of introducing a method to CharSequence for bulk getChars. Doing that would help both here and in Appendable/Writer implementations like StringWriter, PrintWriter, and OutputStreamWriter which currently convert to a String to then write.

https://github.com/openjdk/jdk/blob/master/src/java.base/share/classes/java/io/Writer.java#L367

liach replied:
I believe this batch operation would be a useful addition, but it is best reserved for another RFE.

mkarg replied:
I just filed the PR to implement `CharSequence.getChars` bulk-read method. Please see https://github.com/openjdk/jdk/pull/21730. 🙂

ExE-Boss commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,124 @@ public void close() {
+    public static Reader of(CharSequence source) {
+        Objects.requireNonNull(source);
+
+        return new Reader() {
```
Maybe make this into a package‑private `CharSequenceReader` so that [`StringReader​::close()`] doesn’t have to use a try‑catch.

[`StringReader​::close()`]: https://docs.oracle.com/en/java/javase/23/docs/api/java.base/java/io/StringReader.html#close()

liach replied:
I don't think that try-catch is too much of a bloat compared to the extra license header and imports from the explicit `CharSequenceReader.java` file.

AlanBateman commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,125 @@ public void close() {
+     * calling the {@code close()} method. Subsequent calls to {@code close()}
+     * have no effect.
+     *
+     * <p> After the reader has been closed, the {@code read()},
```
The API docs in 5cbc0450 looks okay. One minor comment is that "After the reader has been closed ..." doesn't need to be a new paragraph as it's part of the lifecycle topic in the previous paragraph.

mkarg replied:
IIUC I shall just remove the `<p>` but keep the text `After the reader...`?

liach replied:
Yep. Better keep everything about closing in the same paragraph.

mkarg commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,125 @@ public void close() {
+        Objects.requireNonNull(cs);
+
+        return new Reader() {
+            private final int length = cs.length();
```
As the anonymous class MUST NOT be used with be used with multiple threads, I always have seen the `CharSequence` as *fixed/static* text in the moment the `Reader` is getting used. But indeed, technically one could interleave `Reader::read()` invocations by `CharSequence.append()` (or even worse, `CharSequence.delete()`) invocations. The question is: Would that make *any* sense in the end? I mean, what happens if one has `read()` text that in the next step gets `delete()`'d? I cannot image *any* scenario where such a program would result in *useful* outcome.

&lt;fun>The fact that nobody so far (before you) brought up this question seems to proof that nobody (besides you) would write such a program. 😄 &lt;/fun>

So I would plea for clearly saying in the JavaDocs that `cs` MUST NOT be modified before `close()` is called. Every other solution implies strange side effects and slower and error-prone implementation of both, anoynous reader *and* test.

@AlanBateman WDYT?

liach replied:
I would treat this specific scenario as one of the "no concurrent usage" examples. Note that by this principle, mutable objects like `StringBuilder` should not override object comparison methods as these states can change, but they do :(

jaikiran replied:
Consider the example of `StringBuffer`, which is a `CharSequence`. Wouldn't something like the following be a logical use of `Reader.of(CharSequence)`, where you create a `Reader` for the underlying sequence (which is not yet populated) and then keep reading through the `Reader` until the underlying sequence's data is finished?

```java
final StringBuffer sb = new StringBuffer();
try (final Reader reader = Reader.of(sb)) {
    final ContentGenerator cg = new ContentGenerator(sb);
    Thread.ofPlatform().start(cg);
    int numRead = 0;
    while (numRead != -1) {
        final char[] content = new char[1024];
        numRead = reader.read(content); // wait for content
    }
}

private class ContentGenerator {
    final StringBuffer sb;
    ContentGenerator(StringBuffer sb) {
        this.sb = sb;
    }

    @Override
    run() {
        while (true) {
            String foo = getSomeContent(); // maybe blocking
            sb.append(foo);
        }
    }
}
```

jaikiran replied:
>  I cannot image any scenario where such a program would result in useful outcome.

An additional point of reference is the current default implementation of `public default IntStream chars()`, which returns an IntStream of char values from the sequence and the implementation of which doesn't consider the `length()` to be fixed.

liach replied:
For `chars()` or `codePoints()`, I believe calling `length()` or not was a matter of implementation convenience instead of the assumption that `length()` can change during calls. Note implementation methods in the anonymous class in `codePoints()` cache the length in local variables. Maybe they just don't want the extra field overhead in the objects they construct.

AlanBateman replied:
> @AlanBateman WDYT?

It's good question as a CharSequence's length can change over time, e.g. StringBuilder. This scenario comes up regularly with InputStreams and Readers connected to files as the file may be growing and shrinking as bytes are read. It would not be surprising to read to EOF/-1 and to not read any further chars, even if the underlying sequence has grow.

AlanBateman replied:
The sentence "If the sequence is modified .." in the these methods is in the context of the methods. For ofReader(CharSequence) then it is any modification, doesn't have to be concurrent. Maybe something like "If the sequence changes while the reader is open, e.g. the length changes, the behavior is undefined".

mkarg replied:
Thank you, Alan. I have copied your wording.

jaikiran commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,125 @@ public void close() {
+            /** Check to make sure that the stream has not been closed */
+            private void ensureOpen() throws IOException {
+                if (isClosed)
+                    throw new IOException("Stream closed");
```
Should the exception text instead be "Reader closed" and also the code comment a couple of lines above, replace "stream" to "reader"?

mkarg replied:
The complete implementation was copied from "StringReader", so it was "Stream closed" there already, and "StringReader" now refers to this new code location. I deliberately did not modify anything else but the essentially needed parts. If we change this text here, this might be unexpected by existing applications.

jaikiran replied:
Okay, it's reasonable to keep it in this form then.

AlanBateman commented:
> @AlanBateman Can you please review [the CSR request](https://bugs.openjdk.org/browse/JDK-8341596) so I can finish it? Thanks!

Latest API docs looks good, will you update the CSR?

mkarg commented:
> > @AlanBateman Can you please review [the CSR request](https://bugs.openjdk.org/browse/JDK-8341596) so I can finish it? Thanks!
> 
> Latest API docs looks good, will you update the CSR?

Done. I have replaced the "Specification" section of the [CSR](https://bugs.openjdk.org/browse/JDK-8341596) a minute ago. If you agree, I will switch the CRS from Draft to Finalized now to start the CSR Review Process. Ok?

jaikiran commented on test/jdk/java/io/Reader/Of.java:
```diff
@@ -0,0 +1,217 @@
+ * @test
+ * @bug 8341566
+ * @run testng Of
+ * @summary Check for expected behavior of Reader.of().
```
Nit - the jtreg documentation recommends the following order of these tags https://openjdk.org/jtreg/tag-spec.html#ORDER. For newly introduced tests, like this one, it will be good to follow that recommendation.

jaikiran commented on test/jdk/java/io/Reader/Of.java:
```diff
@@ -0,0 +1,217 @@
+ * @run testng Of
+ * @summary Check for expected behavior of Reader.of().
+ */
+public class Of {
```
Would it be possible to include a test which creates and operates on a `Reader` which is backed by a `CharSequence` of zero length. Similarly, a test which creates and operates on a `Reader` which is backed by a `CharSequence` of `Integer.MAX_VALUE` length? I think that should give us good coverage for the implementation of the returned `Reader` for these corner cases.

mkarg replied:
Zero-case is partically covered by the existing test code. Your request looks like a real lot of work, as it implies a complete rewrite of the overall structure of the test - and frankly spoken, my budget for this issue is near to spent already. Does that *really* provide a justifying benefit, given the fact that these are literally edge cases (and given the fact, that the code is a 1:1 copy of `StringReader`, which we all used since decades, including those edge cases)? If you think it is really necessary, can you please post the modified version of `Of.java` here? I will then post it into the PR. Thanks.

jaikiran commented on test/jdk/java/io/Reader/Of.java:
```diff
@@ -0,0 +1,217 @@
+                "transferTo() provides wrong content");
+    }
+
+    @Test(dataProvider = "readers", expectedExceptions = IOException.class)
```
Instead of `expectedExceptions` at the method level, I think it would be better to use something like:

```java
org.testng.Assert.assertThrows(IOException.class,
                               () -> {reader.read();});
```
That way it's clear which operation within the test method is expected to throw this exception and it also prevents unexpected `IOException` (for example, from reader.close()) from going unnoticed.

jaikiran commented on test/jdk/java/io/Reader/Of.java:
```diff
@@ -0,0 +1,217 @@
+        reader.read();
+    }
+
+    @Test(dataProvider = "readers", expectedExceptions = IOException.class)
```
Same comment as above, about using assertThrows instead of expectedExceptions at the method level.

jaikiran commented:
In a different discussion, I have been told previously that the title of the CSR and the JBS issue for which the CSR is created, should match. I think we should change the title of https://bugs.openjdk.org/browse/JDK-8341566 to match what you have for the CSR, since that one is more clear.

mkarg commented:
> In a different discussion, I have been told previously that the title of the CSR and the JBS issue for which the CSR is created, should match. I think we should change the title of https://bugs.openjdk.org/browse/JDK-8341566 to match what you have for the CSR, since that one is more clear.

I have no strong feelings, but the reason for deviation is that the JBS proposes to introduce "some" factory (the outcome is open at time of JBS creation and was found _later_, while discussing the PR), while the CSR is explicitly about the exact wish for literally `Reader.of(final CharSequence cs)`. I do not know the intentions of the team that has set up the process rules for JBS and CSR. If both need to be exact the same, then I wonder why JIRA allows to modify the CSR title manually...? The sole answer I could imagine was: Because of exact that difference between "some" factory and exactly `Reader.of`. 🤔 

@AlanBateman WDYT?

AlanBateman commented:
> @AlanBateman WDYT?

A title like "Add Reader.of(CharSequence)" would make it clear to everyone what this issue, and the CSR, is about.

mkarg commented:
> > @AlanBateman WDYT?
> 
> A title like "Add Reader.of(CharSequence)" would make it clear to everyone what this issue, and the CSR, is about.

Changed in PR, RFE and CSR.

mkarg commented:
> > > @AlanBateman Can you please review [the CSR request](https://bugs.openjdk.org/browse/JDK-8341596) so I can finish it? Thanks!
> > 
> > 
> > Latest API docs looks good, will you update the CSR?
> 
> Done. I have replaced the "Specification" section of the [CSR](https://bugs.openjdk.org/browse/JDK-8341596) a minute ago. If you agree, I will switch the CRS from Draft to Finalized now to start the CSR Review Process. Ok?

@AlanBateman Alan, according to https://wiki.openjdk.org/display/csr, "at least one engineer familiar with that technology area" MUST review the CSR before I may update its state from "Draft" to "Finalized". I would be happy if you would be that engineer, so I have just set the assignee of this CSR from me to you. Thanks for your help! :-)

liach commented:
I have looked at the CSR and revised the wording for you, mostly conveying information CSR review wants to see.

The original "Problem" section only mentions performance, but I added consistency with specification and problems with third-party/custom implementations, as those are indeed problems that we are fixing with this new API. Also simplified "Solution" section to emphasize on our new API. The inlined links have been standardized to links in the "Issue Links" section.

@mkarg Can you take a look at this revised version of CSR?

mkarg commented:
> I have looked at the CSR and revised the wording for you, mostly conveying information CSR review wants to see.
> @mkarg Can you take a look at this revised version of CSR?

Thank you, Chen. I have no experience with CSR, so I do trust in yours. Having said that, I now have upgraded the status from "Draft" to "Finalized". Ist that enough to trigger the review process, or do I have to set an assignee / manually trigger somebody?

liach commented:
Wiki pages such as https://wiki.openjdk.org/display/csr/Fields+of+a+CSR+Request are good starting points for CSRs.

CSR review will look at CSRs listed at https://bugs.openjdk.org/secure/Dashboard.jspa?selectPageId=17313. To finalize a CSR for review, you should set your fix version; I fixed that when I marked myself as a reviewer on the CSR.

Also note that the CSR lead might be unavailable the next few days, so the review will wait for a bit. Meanwhile, it's good for other experienced reviewers like Alan to review the CSR too.

RogerRiggs reviewed (commented):
Looks pretty good; some suggestions about clarity of the javadoc.
Claims are made about performance but are minimal and unsubstantiated with JMH tests.
Some updates to the CSR will be needed after updating the javadoc.

RogerRiggs commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,125 @@ public void close() {
     }
 
+    /**
+     * Returns a {@code Reader} that reads characters from a
+     * {@code CharSequence}, starting at the first character in the sequence.
```
I see Alan and I have different views on the inclusion of the details.
With only a single CharSequence argument, I can't see any ambiguity about the starting position.

AlanBateman replied:
> I see Alan and I have different views on the inclusion of the details. With only a single CharSequence argument, I can't see any ambiguity about the starting position.

It could be a second sentence, "Reading starts at the first character in the sequence".

mkarg replied:
Done. Adopted Alan's wording.

liach commented on src/java.base/share/classes/java/io/StringReader.java:
```diff
@@ -30,31 +30,28 @@
  * A character stream whose source is a string.
  *
+ * @apiNote
+ * As of release JDK 24, this class has been supplemented with an equivalent
+ * API designed for use by a single thread, {@link Reader#of(CharSequence)}.
+ * {@code Reader.of(String)} should generally be used in preference to this one,
+ * as it supports all of the same operations but it is faster, as it performs no
+ * synchronization.
```
Currently uncontested locks have a performance regression after the removal of biased locking: A similar example with `StringBuffer` is available in #19513.

mkarg replied:
I do not have any personal preferences here. The text was paraphrased from `StringBuilder`, so it seems it was acceptable in the past. If a different text is wanted, I would kindly ask all participants in this discussion to agree upon one common proposal, as I personally am fine with the text as-is.

mkarg replied:
Done; replaced by Alan's proposal: "Reader.of(CharSequence) provides a method to read from any CharSequence that may be more efficient than StringReader"

mkarg commented on test/jdk/java/io/Reader/Of.java:
```diff
@@ -0,0 +1,217 @@
+ * @test
+ * @bug 8341566
+ * @summary Check for expected behavior of Reader.of().
+ * @run testng Of
```
The test name follows the conventions currently applied in this folder. Feel free to propose a better name, I would be happy to change it.

RogerRiggs replied:
The test name should provide enough context so the *human* reader does not have to hunt around for the context.
Either CharSequenceReader or ReaderOfCharSequence would be easier to understand without having to *know* the conventions.

mkarg replied:
I personally did not have any problem understanding the existing conventions within a second, but OTOH I have no strong feelings here.

@AlanBateman Which of both proposed names shall I use (or shall I keep the current name)?

AlanBateman commented on src/java.base/share/classes/java/io/StringReader.java:
```diff
@@ -30,31 +30,28 @@
+ * API designed for use by a single thread, {@link Reader#of(CharSequence)}.
+ * {@code Reader.of(String)} should generally be used in preference to this one,
+ * as it supports all of the same operations but it is faster, as it performs no
+ * synchronization.
```
We shouldn't using "release 24" or "supplemented" here. The API note just needs to `Reader.of(CharSequence)` provides a method to read from any CharSequence that may be more efficient than `StringReader`.

liach replied:
I think "release 24" is fine for now as similar text exists in `java.util.Vector`. We can normalize such texts when we implement https://bugs.openjdk.org/browse/JDK-8341736.

RogerRiggs replied:
Please remove the reference to JDK 24; it will age rapidly and would require extra maintenance later.
The javadoc should focus on the functionality provided.

mkarg replied:
Done; replaced by Alan's proposal: "Reader.of(CharSequence) provides a method to read from any CharSequence that may be more efficient than StringReader"

mkarg commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,125 @@ public void close() {
+     * {@code read(CharBuffer)}, {@code ready()}, {@code skip(long)}, and
+     * {@code transferTo()} methods all throw {@code IOException}.
+     *
+     * <p> The returned reader supports the {@link #mark mark()} operation.
```
I share both visions, so I have adopted #reset now as it does not harm IMHO.

mkarg commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,125 @@ public void close() {
+     * @return a {@code Reader} which reads characters from {@code cs}
+     *
+     * @throws NullPointerException if {@code cs} is {@code null}
+     *
```
done

ecki commented:
> I have slightly adapted the copy loop to allow CPUs to use solely single-operand INC within the loop instead of dual-operand ADD.

If that makes a difference we might need to consider fixing the Compiler :)

mkarg commented:
> > I have slightly adapted the copy loop to allow CPUs to use solely single-operand INC within the loop instead of dual-operand ADD.
> 
> If that makes a difference we might need to consider fixing the Compiler :)

Hm... really? In fact I am not sure if the compiler is actually able to optimize replacement from ADD to INC *in this particular* scenario: The operand of the ADD is *neither* a constant and in particular *not* `1`, the upper limit of the loop is *not* a constant, and is typically far beyond any reasonable unrolling limit (readers rather often are asked for multiple KB here to optimize I/O transmissions). This is not just `for (int i = 0; i < 5; i++;)` (always same limit, reasonably low limit, using INC only).

Anyways, if you can proof me wrong I am willing to learn (and to undo this commit)! :-)

RogerRiggs commented:
> I have slightly adapted the copy loop to allow CPUs to use solely single-operand INC within the loop instead of dual-operand ADD.

Did you measure a performance improvement? Compiler optimizations are more productive than spot source changes.

mkarg commented:
> > I have slightly adapted the copy loop to allow CPUs to use solely single-operand INC within the loop instead of dual-operand ADD.
> 
> Did you measure a performance improvement? Compiler optimizations are more productive than spot source changes.

No and I do not intent to do, as it was ja a code simplification I have seen, not an engineering target actively developed to reach max performance. Did you measure a performance drawback of the latest code? Maybe Compiler optimization is less productive in that particular case, due to the constraints mentioned?

RogerRiggs commented:
That's fine, but talking about performance improvements because of what you think the compiler will do isn't productive.
Make the source clear and maintainable and low maintenance.  tnx

mkarg commented:
> That's fine, but talking about performance improvements because of what you think the compiler will do isn't productive. Make the source clear and maintainable and low maintenance. tnx

Uhm... actually I thought my change *made* it clear and maintainable, as now the loop runs from start to end instead from zero to count 🤔. Anyways, I have no strong feelings here. If you like it more to iterate from zero to count then I simply can undo the commit (even if I personally think that is *less* understandable).

mkarg commented:
@liach Just to make it clear: Thumbs up for "current code *is* more understandible" or thumbs up for "undo last commit"? 🤔

RogerRiggs commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,127 @@ public void close() {
+                    case StringBuffer sb -> sb.getChars(next, next + n, cbuf, off);
+                    case CharBuffer cb -> cb.get(next, cbuf, off, n);
+                    default -> {
+                        for (int i = next, j = next + n; i < j;)
+                            cbuf[off++] = cs.charAt(i++);
```
The single increment is fine and readable. 
Introducing a local variable for the end (j) seems unnecessary and its name is not indicative of its purpose as a limit.

liach replied:
I would say this current version is fine: that's how we handle encode/decode loop in classes like `String`.

jaikiran replied:
I agree with Roger - the previous version you had:

```java
for (int i = 0; i < n; i++)
    cbuf[off + i] = cs.charAt(next + i);

```
was trivial, easy to understand immediately without having to stare at it for a while to see what's being done there, compared to the current version:

```java
for (int i = next, j = next + n; i < j;)
    cbuf[off++] = cs.charAt(i++);
```
I prefer switching to the former version.

mkarg replied:
switched to previous version to end discussion

liach commented:
> @liach Just to make it clear: Thumbs up for "current code _is_ more understandible" or thumbs up for "undo last commit"? 🤔

My thumb up is for this:

> Anyways, I have no strong feelings here

This debate is really pointless bikeshedding. Either way is fine to me.

mkarg commented:
> ... that's how we handle encode/decode loop in classes like `String` ... This debate is really pointless bikeshedding. Either way is fine to me.

Chen, I completely share your opinion here, but as Roger and Jaikiran apparently do have string feelings here, I have simply undone this last commit. This detail discussion is not worth spending any more of everybody's precious time.

jaikiran commented:
Thank you for the updates related to the review comments. I don't have anything more.
My only other comment was test related - to additionally test the returned `Reader`'s APIs against corner cases like an empty `CharSequence` and a `CharSequence` with `Integer.MAX_VALUE` length. However, if you have run out of time and energy on this one, then the test is fine in the current form. I don't have the necessary bandwidth to propose a test snippet for it right now and we can enhance the test separately at a later date.

jaikiran reviewed (approved):
The latest version of this PR (commit `487b9386`) looks OK to me.

jaikiran commented on test/jdk/java/io/Reader/Of.java:
```diff
@@ -0,0 +1,217 @@
+    }
+
+    @Test(dataProvider = "readers")
+    public void testReadBII(Reader reader) throws IOException {
```
What does `BII` in some of these test method names represent?

mkarg replied:
Good question. I hereby forward it to the original author of the existing tests I have copied from OpenJDK. I assume it was some kind of convention.

eirbjo commented on src/java.base/share/classes/java/io/Reader.java:
```diff
@@ -140,6 +140,127 @@ public void close() {
+                if (next >= length)
+                    return -1;
+                int n = Math.min(length - next, len);
+                switch (cs) {
+                    case String s -> s.getChars(next, next + n, cbuf, off);
+                    case StringBuilder sb -> sb.getChars(next, next + n, cbuf, off);
+                    case StringBuffer sb -> sb.getChars(next, next + n, cbuf, off);
+                    case CharBuffer cb -> cb.get(next, cbuf, off, n);
+                    default -> {
+                        for (int i = 0; i < n; i++)
+                            cbuf[off + i] = cs.charAt(next + i);
+                    }
+                }
```
Just a drive-by observation, not for this PR to handle:

If four out of five `CharSequence` implementations in the JDK include fast-path methods for copying char ranges into an output buffer, perhaps it is time to introduce `CharSequence::getChars`` as a default method? 

That way, code like this wouldn't need to special-case for the known fast-paths, and external implementations could also join the party.

robtimus replied:
That had been discussed here: https://github.com/openjdk/jdk/pull/21371#pullrequestreview-2350588453

mkarg replied:
Stay tuned. I have already forwarded a subsequent RFE and PR to cover this topic. 🤩

mkarg replied:
I just filed the PR to implement `CharSequence.getChars` bulk-read method. Please see https://github.com/openjdk/jdk/pull/21730. 🙂

RogerRiggs reviewed (approved):
Thanks for the updates; LGTM

mkarg commented:
I have aligned this PR with the latest CSR proposal. Kindly asking for reviews. Thanks.

liach reviewed (approved):
The latest revision looks good.

Since I am not a professional reviewer in the IO area, I will require extra reviewer approvals.

/reviewers 2 reviewer

jmehrens commented:
Will the Reader need to catch IndexOutOfBoundsException and BufferUnderFlowException then wrap them as EOFException? I would think these errors should not escape the Reader as is if a source CharBuffer limit is tampered with.  Or is this not an issue?

liach commented:
@jmehrens We have this in our specification:

>  If the sequence changes while the reader is open, e.g. the length changes, the behavior is undefined.

Thus your scenario is not an issue.

mkarg commented:
@AlanBateman As the CSR is approved, kindly asking for your review of this PR to finally integrate the changes. Thanks.

mkarg commented:
@parttimenerd FYI :-)

AlanBateman commented:
> @AlanBateman As the CSR is approved, kindly asking for your review of this PR to finally integrate the changes. Thanks.

I suggested this factory method so my interest/focus has been in its specification, I have not had cycles to look at the implementation or tests but you've got plenty of eyes on it. Minor edits would be make "The reader is initially .." the second sentence of the first paragraph rather than a new paragraph. Also "The returned reader supports the mark and reset ..." could move up to become the second paragraph. Otherwise happy that this ended up a static factory method rather than a new implementation class in the API.

mkarg commented:
@liach Alan's latest minor edit request auto-removed your approval from this PR. Kindly asking to reapply your approval. Thanks.

liach commented:
The javadoc text changes look fine; since this does not constitute specification changes, we don't need to re-draft our CSR to reflect the text changes.

liach reviewed (approved):
The diff from Jaikiran and Roger's latest review is: https://github.com/openjdk/jdk/compare/487b93864274f5a27449e8ce0d925be333db6600...97e3144cde798090a113d5cd3176cbc7fb3ef759

There's no implementation change, so assuming the previous reviewers' approvals hold.

mkarg commented:
I do share Chen's assumption, therefore requesting integration now.

/integrate

jaikiran reviewed (approved):
This looks good to me.

kevinb9n commented:
An issue with the simple name `of` is that we may encounter calls like this:

`Reader.of("/usr/share/dict/words")`

... that may look like they do something different. I can't say whether we should consider this a serious issue or not, though.

liach commented:
I think this scenario has a close analogy with java.util.Scanner constructors. Does such a fault happen to Scanner often?

liach commented:
Thanks Markus! The java.io tests succeed with repeats and the only test failures in tier 1-3 were ones problem-listed by f7a61fce949ea2d15ec09d295c643d0c2eabea1e.

FYI, besides Kevin's question, there was another concern about this API's interaction with BufferedReader (as you know, many Reader users wrap any Reader into a BufferedReader, especially you want to read all lines from a reader). We currently decide that we might explore line reading capability to Reader, but anyways that will not affect this patch.

And I have consulted with Kevin in private; he thinks this minor confusion is fine.

Sponsoring.

/sponsor

mkarg commented:
I have drafted a release note for [JDK-8341566](https://bugs.openjdk.org/browse/JDK-8341566) in [JDK-8344910](https://bugs.openjdk.org/browse/JDK-8344910). Kindly asking for reviews! :-)

AlanBateman commented:
> I have drafted a release note for [JDK-8341566](https://bugs.openjdk.org/browse/JDK-8341566) in [JDK-8344910](https://bugs.openjdk.org/browse/JDK-8344910). Kindly asking for reviews! :-)

Thanks for creating this. I've edited it so that it is now focused on highlighting the new API and nothing else. Chen had some good suggestions too.
